File: Prob_004/prob_004.py
#!/usr/bin/env python3
#
# Project Euler.net Problem 4
#
# A palindromic number reads the same both ways.
#
# The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 x 99.
# Find the largest palindrome made from the product of two 3-digit numbers.


def palindrome(candidate_int):
    # Convert the candidate number to a string
    candidate_str = "{}".format(candidate_int)
    candidate_str_rev = candidate_str[::-1]

    # Check if it is a palindrome
    return candidate_str == candidate_str_rev


max_palindrome = 0
max_i = max_j = 0
for i in range(999, 99, -1):
    for j in range(999, 99, -1):
        candidate = i * j
        if ((candidate > max_palindrome) and palindrome(candidate)):
            max_palindrome = candidate;
            (max_i, max_j) = (i, j)
            print("i = {}, j = {}, candidate = {}".format(i, j, candidate))


print("Largest palindrome is {} from multiplying {} by {}.".format(max_palindrome, max_i, max_j))

File: Prob_004/prob_004.scala
// Project Euler.net Problem 4
//
// A palindromic number reads the same both ways.
//
// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 x 99.
// Find the largest palindrome made from the product of two 3-digit numbers.


def palindrome (candidate_int: Int): Boolean = {
    val candidate_str = candidate_int.toString
    val length = candidate_str.length
    for (i <- 0 until length) {
        if (candidate_str(i) != candidate_str(length-1-i))
            return false
    }
    true
}

var max_palindrome = 0
var max_i, max_j = 0

for (i <- 100 to 999; j <- i to 999) {
    val candidate = i * j
    if ((candidate > max_palindrome) && palindrome(candidate)) {
        max_palindrome = candidate
        max_i = i
        max_j = j
    }
}
println("Largest palindrome is " + max_palindrome + " from multiplying " + max_i + " by " + max_j + ".")
